src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,1109 @@
+            return;
+        }
+
+        let chains = std::mem::take(&mut self.chains);
+        let (merged, remains) = merge_chains_up_to(chains, &self.since, upper);
+
+        self.chains = remains;
+        if !merged.is_empty() {
+            // We put the merged chain at the end, assuming that its contents are likely to
+            // consolidate with retractions that will arrive soon.
+            self.chains.push(merged);
+        }
+
+        // Restore the chain invariant.
+        //
+        // This part isn't great. We've taken great care so far to only look at updates with times
+        // before `upper`, but now we might end up merging all chains anyway in the worst case.
+        // There might be something smarter we could do to avoid merging as much as possible. For
+        // example, we could consider sorting chains by length first, or inspect the contained
+        // times and prefer merging chains that have a chance at consolidating with one another.
+        let mut i = self.chains.len().saturating_sub(1);
+        while i > 0 {
+            let needs_merge = self.chains.get(i).is_some_and(|a| {
+                let b = &self.chains[i - 1];
+                a.len() * CHAIN_PROPORTIONALITY > b.len()
+            });
+            if needs_merge {
+                let a = self.chains.remove(i);
+                let b = std::mem::take(&mut self.chains[i - 1]);
+                let merged = merge_chains(vec![a, b], &self.since);
+                self.chains[i - 1] = merged;
+            } else {
+                i -= 1;
+            }
+        }
+
+        self.update_metrics();
+    }
+
+    /// Advance the since frontier.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given `since` is less than the current since frontier.
+    pub fn advance_since(&mut self, since: Antichain<Timestamp>) {
+        assert!(PartialOrder::less_equal(&self.since, &since));
+        self.since = since;
+    }
+
+    /// Consolidate all updates at the current `since`.
+    pub fn consolidate_at_since(&mut self) {
+        let upper_ts = self.since.as_option().and_then(|t| t.try_step_forward());
+        if let Some(upper_ts) = upper_ts {
+            let upper = Antichain::from_elem(upper_ts);
+            self.consolidate_before(&upper);
+        }
+    }
+
+    /// Update persist sink metrics.
+    fn update_metrics(&mut self) {
+        let mut new_size = LengthAndCapacity::default();
+        for chain in &mut self.chains {
+            new_size += chain.get_size();
+        }
+
+        let old_size = self.total_size;
+        let len_delta = UpdateDelta::new(new_size.length, old_size.length);
+        let cap_delta = UpdateDelta::new(new_size.capacity, old_size.capacity);
+        self.metrics
+            .report_correction_update_deltas(len_delta, cap_delta);
+        self.worker_metrics
+            .report_correction_update_totals(new_size.length, new_size.capacity);
+
+        self.total_size = new_size;
+    }
```

Comment:
We should use something similar to the arrangement size logging instead. This only exposes it through metrics. Can fix as a separate PR!

Reply:
Good point! Though this needs some more design work about how to expose these sizes. Exposing them in `mz_arrangement_sizes` would be confusing.

Reply:
Opened https://github.com/MaterializeInc/database-issues/issues/8914 to track.